dtool_lookup_gui/utils/logging.py

```diff
@@ -0,0 +1,217 @@
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+import logging
+from typing import Optional
+
+from gi.repository import Gtk, Pango
+
+DEFAULT_FORMATTER = logging.Formatter('%(asctime)s:%(name)s:%(levelname)s: %(message)s')
+SINGLE_MESSAGE_FORMATTER = logging.Formatter('%(levelname)s: %(message)s')
+
+DEFAULT_TEXT_BUFFER_MAX_LINES = 1000
+DEFAULT_ENTRY_MAX_LINES = 5
+
+
+# formatter mixins
+
+class FormattedHandlerMixin():
+    """Mixin for logging.Handler derivatives. Assigns formatter at creation."""
+    def __init__(self, *args,
+                 formatter: Optional[logging.Formatter] = DEFAULT_FORMATTER,
+                 **kwargs):
+        """Attach formatter to Handler instance."""
+        super().__init__(*args, **kwargs)
+        self.setFormatter(formatter)
+
+
+class SingleMessageFormatHandlerMixin(FormattedHandlerMixin):
+    """Mixin for logging.Handler derivatives. Assigns formatter at creation."""
+
+    def __init__(self, *args,
+                 formatter: Optional[logging.Formatter] = SINGLE_MESSAGE_FORMATTER,
+                 **kwargs):
+        """Attach formatter to Handler instance."""
+        super().__init__(*args, formatter=formatter, **kwargs)
+
+
+# handlers
+
+class GtkBufferHandler(logging.Handler):
+    def __init__(self, *args,
+                 max_lines: Optional[int] = None,
+                 **kwargs):
+        """Abstract base class for any Gtk buffer handler.
+
+        Limit maximum number of lines if max_lines set to int"""
+        super().__init__(*args, **kwargs)
+        self.max_lines = max_lines
+
+    @property
+    def max_lines(self, value: int):
+        return self._max_lines
+
+    @max_lines.setter
+    def max_lines(self, value: Optional[int] = None):
+        if value is None:
+            self._max_lines = None # no limit
+        elif isinstance(value, int) and value >= 0:
+            self._max_lines = value
+        else:
+            raise ValueError(f"max_lines must be None or int, not {type(value)}.")
+
+    def _insert_into_buffer(msg):
+        ...
```

Comment:
Is this supposed to be an ellipsis?

Reply:
Yes, that's there intentionally as an abstract method.

Reply:
Okay, didn't know this syntax...

Reply:
Okay, didn't know this syntax...